doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+  The original ``resolve_descriptors`` function will be called between these
+  two calls, so that the output descriptors may not be set in the first call.
+
+The ``view_inputs`` method allows passing the correct inputs into the
+original ``resolve_descriptors`` function, while ``wrap_outputs`` ensures
+the correct descriptors are used for output allocation and input buffering casts.
+
+The second use-case can be implemented within the inner-loop and
+``resolve_descriptors`` functions.
+This will incur additional overheads, partially because the loop dispatch
+effectively has to be done twice, but mostly because an efficient
+implementation will require direct access to the original
+setup/teardown, etc. functions.
+Such functionality can be added at a later time.
+
+
+Details for ``resolve_descriptors``
+"""""""""""""""""""""""""""""""""""
+
+*TODO: Should this function also get the full set of information which
+I want to pass in to the setup/teardown/inner-loop?  On the one hand, much
+of the information is not yet available/defined (or is set here).  On the
+other hand, some of the info is useful, and it may be nice to just have
+a homogeneous calling convention.*
+
+The UFunc machinery must know the correct dtypes to use before arrays can
+be allocated. The arrays creation itself is desirable to happen before any
+setup functionality to allow potential choosing of an optimized loops.
+
+**Notes:**
+
+There are a few possible extension to this function.  Currently, it also
+takes care of casting in general.  This is not necessary, however, it
+would be possible to extend the signature with casting indication for
+*outputs*.
+This would allow registering e.g. ``float64 + float64 -> float32`` as an explicit
+(faster) loop while indicating that it is an unsafe cast on the result array,
+which requires the user to specify ``casting="unsafe"``.
+
+The current design allows such a specialized loop (with access to the
+initially private ``setup``), from within the ``float64+float64->float64``
+implementation only.
+
+
+Bound UFuncImpl
+"""""""""""""""
+
+The ``UFuncImpl`` is much like a bound method on an object, which in Python
+are passed the ``self`` argument.  There are some small difference in that
+a ``UFuncImpl`` represents a set of such bound methods.
+
+With respect to a ``UFuncImpl``, the normally passed ``self`` argument
+represents the following set of information:
+
+* The ufunc itself, which may be useful for information such as the ufunc's name
+  but is also required to match up information such as the ``ufunc.signature``.
+* The DTypes involved, which are thus passed into all of the methods of the
+  ``UFuncImpl`` in some form or another.
+
+Both of these are *not* stored on the ``UFuncImpl`` itself and just like
+methods, most of the time ``UFuncImpl`` should only be used indirectly by
+calling the ufunc it is bound to.
+It is possible to bind a single ``UFuncImpl`` to multiple ufuncs.
```

Comment:
I think this is crucial and deserves mention early on! Here, a specific example might be good. E.g., I think `sin` and `cos` could share complete implementations, but `add` and `subtract` could not (e.g., `add` being defined for strings but not for `datetime`, and vice versa for `substract`).